python/cuml/pyproject.toml

```diff
@@ -19,6 +19,14 @@ requires = [
     "scikit-build-core[pyproject]>=0.10.0",
 ] # This list was generated by `rapids-dependency-file-generator`. To make changes, edit ../../dependencies.yaml and run `rapids-dependency-file-generator`.
 
+[tool.pydistcheck]
+select = [
+    "distro-too-large-compressed",
+]
+
+# detect when package size grows significantly
+max_allowed_size_compressed = '1.5G'
```

Comment:
Is this value coming from the existing wheel size, or coming from somewhere else?

Reply:
This seems too big. cuML wheels appear to be closer to 550MB. Source: https://anaconda.org/rapidsai-wheels-nightly/cuml-cu12/files

Maybe set the threshold at 600MB.

Reply:
This is the existing wheel size + a buffer. It varies by CPU architecture, Python version (because of Cython stuff), and CUDA version (because, for example, we don't use CUDA math lib wheels for CUDA 11).

The largest one I've seen clicking through logs on this PR was `CUDA 11.8.0, Python 3.10, amd64`:

```text
checking 'final_dist/cuml_cu11-24.12.0a38-cp310-cp310-manylinux_2_28_x86_64.whl'
----- package inspection summary -----
file size
  * compressed size: 1.3G
  * uncompressed size: 2.2G
  * compression space saving: 42.2%
contents
  * directories: 72
  * files: 432 (86 compiled)
size by extension
  * .so - 2.2G (99.9%)
  * .py - 1.4M (0.1%)
  * .pyx - 1.2M (0.1%)
  * .0 - 0.2M (0.0%)
  * .ipynb - 0.1M (0.0%)
  * no-extension - 57.2K (0.0%)
  * .png - 51.3K (0.0%)
  * .pxd - 34.0K (0.0%)
  * .txt - 25.7K (0.0%)
  * .md - 10.3K (0.0%)
  * .h - 2.1K (0.0%)
  * .ini - 0.8K (0.0%)
largest files
  * (2.2G) cuml/libcuml++.so
  * (3.0M) cuml/experimental/fil/fil.cpython-310-x86_64-linux-gnu.so
  * (2.9M) cuml/fil/fil.cpython-310-x86_64-linux-gnu.so
  * (1.5M) cuml/cluster/hdbscan/hdbscan.cpython-310-x86_64-linux-gnu.so
  * (1.5M) cuml/svm/linear.cpython-310-x86_64-linux-gnu.so
------------ check results -----------
errors found while checking: 0
```

([build link](https://github.com/rapidsai/cuml/actions/runs/11821309838/job/32946567266?pr=6136#step:13:2120))

So proposing setting this to around 200MB above that size, so we'd be notified if the binary size increased above that level.

There's nothing special about 1.5GB... it's already way way too big to be on PyPI. But proposing putting *some* limit so that we can get automated feedback from CI about binary size growth, and make informed decisions about whether to do something about it... similar to setting a coverage threshold for tests.

Reply:
Aaahhh, but CUDA 11 is huge. We only did CUDA wheels work for CUDA 12. https://anaconda.org/rapidsai-wheels-nightly/cuml-cu11/files

Reply:
Yeah exactly:

https://github.com/rapidsai/cuml/blob/56e5e622a950efce365562d8f17bb38c3608d980/python/cuml/CMakeLists.txt#L96-L104

Reply:
@bdice re https://github.com/rapidsai/cuml/pull/6136#pullrequestreview-2434724069

> *Maybe we can make it depend on CUDA version? Or just wait until we drop CUDA 11.*

Yeah totally. The specific implementation I can think of would look like removing this configuration from `pyproject.toml` and instead putting it into the script, like this (pseudocode):

```shell
if [[ ${CUDA_MAJOR_VERSION} == "11" ]]; then
    pydistcheck \
        --inspect \
        --select 'distro-too-large-compressed' \
        --max-allowed-size-compressed '1.5G' \
        final_dist/*.whl
else
    pydistcheck \
        --inspect \
        --select 'distro-too-large-compressed' \
        --max-allowed-size-compressed '750M' \
        final_dist/*.whl
fi
```

Want me to do that? I'm indifferent about it... on the one hand it will help us catch specific-to-one-CUDA-version issues, on the other hand it makes the scripts a bit more complex and a bit different from most other RAPIDS repos. Happy to defer to whichever you want.

Reply:
I’m indifferent as well. Let’s stick to the single definition for now.

Reply:
alright sounds good, thanks for considering it. I'm glad these changes are helping to expose these differences and leading to these conversations 😊

Reply:
I'd be a fan having two different limits. Mostly because for "not CUDA 11" the limit of 1.5GB might as well be "infinity". As in, if we ever reach it, it will be way to late to course correct.

Should I make a PR that uses Jams' suggestion for two limits?

Reply:
@betatim I've put up PRs in other repos following this suggestion, if you'd like something to copy from here in `cuml`:

* https://github.com/rapidsai/cuvs/pull/469
* https://github.com/rapidsai/cugraph/pull/4754